middle_end/flambda/types/env/aliases.ml

```diff
@@ -128,43 +147,51 @@ end = struct
       Name_mode.Map.merge (fun _order elts1 elts2 ->
           match elts1, elts2 with
           | None, None | Some _, None | None, Some _ -> None
-          | Some elts1, Some elts2 -> Some (Name.Set.inter elts1 elts2))
+          | Some elts1, Some elts2 -> Some (map_inter elts1 elts2))
         t1.aliases t2.aliases
     in
     let t =
       { aliases;
-        all = Name.Set.inter t1.all t2.all;
+        all = map_inter t1.all t2.all;
       }
     in
     invariant t;
     t
 
   let rename rename_name { aliases; all } =
-    let aliases =
-      Name_mode.Map.map (fun elts -> Name.Set.map rename_name elts)
-        aliases
+    let map_name elts =
+      Name.Map.fold (fun elt coercion acc ->
+        Name.Map.add (rename_name elt) coercion acc)
+        elts
+        Name.Map.empty
     in
-    let all = Name.Set.map rename_name all in
-    { aliases; all }
+    let aliases = Name_mode.Map.map map_name aliases in
+    let all = map_name all in
+    let t = { aliases; all } in
+    invariant t; (* CR xclerc for xclerc: not guaranteed to hold *)
```

Comment:
It looks like something needs fixing here!

Reply:
In any event, Xavier confirms that the only way the invariant wouldn't hold is if the two arguments mapped the same name to different name modes, so I added documentation for `Aliases.merge` spelling out this assumption.